src/rust/src/x509/verify.rs

```diff
@@ -0,0 +1,218 @@
+    fn time<'p>(&self, py: pyo3::Python<'p>) -> pyo3::PyResult<&'p pyo3::PyAny> {
+        datetime_to_py(py, self.0.time())
+    }
+}
+
+#[pyo3::prelude::pyfunction]
+fn create_policy(
+    py: pyo3::Python<'_>,
+    profile: &pyo3::PyAny,
+    name: Option<&pyo3::types::PyString>,
+    time: Option<&pyo3::PyAny>,
+) -> pyo3::PyResult<PyPolicy> {
+    let x509_module = py.import(pyo3::intern!(py, "cryptography.x509"))?;
+    let rfc5280 = x509_module
+        .getattr(pyo3::intern!(py, "Profile"))?
+        .get_item(pyo3::intern!(py, "RFC5280"))?;
+    let webpki = x509_module
+        .getattr(pyo3::intern!(py, "Profile"))?
+        .get_item(pyo3::intern!(py, "WebPKI"))?;
+
+    let name = match name {
+        Some(name) => Some(
+            DNSString::new(name.to_str()?)
+                .ok_or_else(|| pyo3::exceptions::PyValueError::new_err("invalid DNS name"))?,
+        ),
+        None => None,
+    };
+
+    let time = time.map(|time| py_to_datetime(py, time)).transpose()?;
+
+    if profile.eq(rfc5280)? {
+        Ok(PyPolicy(FixedPolicy::RFC5280(Policy::new(name, time))))
+    } else if profile.eq(webpki)? {
+        Ok(PyPolicy(FixedPolicy::WebPKI(Policy::new(name, time))))
+    } else {
+        Err(pyo3::exceptions::PyValueError::new_err(
+            "invalid profile specified; expected a Profile variant",
+        ))
+    }
+}
+
+#[pyo3::prelude::pyfunction]
+// #[pyo3(signature = (leaf, policy, intermediates, store))]
+fn verify<'p>(
+    py: pyo3::Python<'p>,
+    leaf: &PyCertificate,
+    policy: &PyPolicy,
+    intermediates: &'p pyo3::types::PyList,
+    store: &'p pyo3::PyAny,
+) -> CryptographyResult<Vec<PyCertificate>> {
+    let intermediates = intermediates
+        .iter()
+        .map(|o| o.extract::<pyo3::PyRef<'p, PyCertificate>>())
+        .collect::<Result<Vec<_>, _>>()?;
+    let store_certs = store
+        .getattr(pyo3::intern!(py, "_certs"))?
+        .downcast::<pyo3::types::PyList>()?
+        .iter()
+        .map(|o| o.extract::<pyo3::PyRef<'p, PyCertificate>>())
+        .collect::<Result<Vec<_>, _>>()?;
+    let store = Store::new(store_certs.iter().map(|t| t.raw.borrow_dependent().clone()));
+
+    // TODO: This manual monomorphization over each policy type is pretty ugly;
+    // maybe fix it.
+    let chain = match &policy.0 {
+        FixedPolicy::RFC5280(policy) => cryptography_x509_validation::verify(
+            leaf.raw.borrow_dependent(),
+            intermediates
+                .iter()
+                .map(|i| i.raw.borrow_dependent().clone()),
+            policy,
+            &store,
+        ),
+        FixedPolicy::WebPKI(policy) => cryptography_x509_validation::verify(
+            leaf.raw.borrow_dependent(),
+            intermediates
+                .iter()
+                .map(|i| i.raw.borrow_dependent().clone()),
+            policy,
+            &store,
+        ),
+    }
+    .map_err(|e| pyo3::exceptions::PyValueError::new_err("validation failed: {e}"))?;
```

Comment:
Flagging for review: this is a lot of duplication (which I can reduce slightly), but it also avoids any `dyn` trait objects being pushed down into the verification API.